tensorflow/python/ops/control_flow_ops.py

```diff
@@ -1917,6 +1919,64 @@ def f2(): return tf.add(y, 23)
 # pylint: enable=g-doc-args
 
 
+def smart_cond(pred, true_fn=None, false_fn=None, name=None):
+  """Return either `true_fn()` if predicate `pred` is true else `false_fn()`.
+
+  If `pred` is a bool or has a constant value, we return either `true_fn()`
+  or `false_fn()`, otherwise we use `tf.cond` to dynamically route to both.
+
+  Arguments:
+    pred: A scalar determining whether to return the result of `true_fn` or
+      `false_fn`.
+    true_fn: The callable to be performed if pred is true.
+    false_fn: The callable to be performed if pred is false.
+    name: Optional name prefix when using `tf.cond`.
+
+  Returns:
+    Tensors returned by the call to either `true_fn` or `false_fn`.
+
+  Raises:
+    TypeError: If `true_fn` or `false_fn` is not callable.
+  """
+  if not callable(true_fn):
+    raise TypeError('`true_fn` must be callable.')
+  if not callable(false_fn):
+    raise TypeError('`false_fn` must be callable.')
+
+  pred_value = constant_value(pred)
+  if pred_value is not None:
+    if pred_value:
+      return true_fn()
+    else:
+      return false_fn()
+  else:
+    return cond(pred, true_fn=true_fn, false_fn=false_fn, name=name)
+
+
+def constant_value(pred):
+  """Return the bool value for `pred`, or None if `pred` had a dynamic value.
+
+  Arguments:
+    pred: A scalar, either a Python bool or a TensorFlow boolean variable
+      or tensor.
+
+  Returns:
+    True or False if `pred` has a constant boolean value, None otherwise.
+
+  Raises:
+    TypeError: If `pred` is not a Variable, Tensor or bool.
+  """
+  if isinstance(pred, bool):
+    pred_value = pred
+  elif isinstance(pred, variables.Variable):
+    pred_value = None
+  elif isinstance(pred, ops.Tensor):
```

Comment:
call this smart_constant_value and add it to tf.contrib.framework?

Reply:
Yeah, I agree with you that the name is better and will submit a commit later.
I also considered whether to put this function in core or not and finally I found that the function was already used as internal interface in the core module before, I mean it's in `tensorflow/python/layers/utils.py` and used by some other modules like `python/layers/normalization.py`. What I did here should just make it public and reusable, so I didn't move it out of the core framework.

Reply:
It can continue to sit here, but for the public API expose it in contrib.  The core API has much stricter behavior requirements and functions cannot be easily modified once exposed in core.

Reply:
I get to know what you mean and I agree. I tried to submit a commit for that.